libbeat/cloudid/cloudid.go

```diff
@@ -0,0 +1,117 @@
+// package cloudid contains functions for parsing the cloud.id and cloud.auth
+// settings and modifying the configuration to take them into account.
+package cloudid
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// OverwriteSettings modifies the received config object by overwriting the
+// output.elasticsearch.hosts, output.elasticsearch.username, output.elasticsearch.password,
+// setup.kibana.host settings based on values derived from the cloud.id and cloud.auth
+// settings.
+func OverwriteSettings(cfg *common.Config) error {
+
+	cloudID, _ := cfg.String("cloud.id", -1)
+	cloudAuth, _ := cfg.String("cloud.auth", -1)
+
+	if cloudID == "" && cloudAuth == "" {
+		// nothing to hack
+		return nil
+	}
+
+	logp.Debug("cloudid", "cloud.id: %s, cloud.auth: %s", cloudID, cloudAuth)
+	if cloudID == "" {
+		return errors.New("cloud.auth specified but cloud.id is empty. Please specify both.")
+	}
+
+	// cloudID overwrites
+	esURL, kibanaURL, err := decodeCloudID(cloudID)
+	if err != nil {
+		return errors.Errorf("Error decoding cloud.id: %v", err)
+	}
+
+	logp.Info("Setting Elasticsearch and Kibana URLs based on the cloud id: output.elasticsearch.hosts=%s and setup.kibana.host=%s", esURL, kibanaURL)
+
+	esURLConfig, err := common.NewConfigFrom([]string{esURL})
+	if err != nil {
+		return err
+	}
+
+	err = cfg.SetChild("output.elasticsearch.hosts", -1, esURLConfig)
```

Comment:
before setting anything in `output.elasticsearch.X`, check the config is actually enabled/available via:

```
esConfig, _ := cfg.GetChild("output.elasticsearch")
esEnabled = esConfig.Enabled()

kibConfig, _ := cfg.GetChild("setup.kibana")
kibEnabled = kibConfig.Enabled()
```

By setting values on a non-present section, you will automatically enable the section (`enabled` always defaults to true).